src/components/user/ExternalAuth.tsx

```diff
@@ -9,53 +15,156 @@ import { useUnauthorizedApi } from 'hooks/common/useDataApi';
 
-  const { token, handleLogin } = useContext(UserContext);
   const unauthorizedApi = useUnauthorizedApi();
+  const history = useHistory();
 
   const isFirstRun = useRef<boolean>(true);
 
-  const settingsContext = useContext(SettingsContext);
-  const externalAuthLoginUrl = settingsContext.settingsMap.get(
-    SettingsId.EXTERNAL_AUTH_LOGIN_URL
-  )?.settingsValue;
+  const { handleLogin } = useContext(UserContext);
+  const { settingsMap } = useContext(SettingsContext);
+
+  const [View, setView] = React.useState<JSX.Element | null>(null);
 
   useEffect(() => {
     if (!isFirstRun.current) {
       return;
     }
+
     isFirstRun.current = false;
 
-    if (!externalToken) {
-      return;
-    }
+    const ErrorMessage = (props: { message?: string }) => (
+      <CenteredAlert
+        severity="error"
+        action={
+          <Button
+            color="inherit"
+            size="small"
+            variant="outlined"
+            onClick={() => {
+              localStorage.clear();
+              window.location.assign('/');
+            }}
+          >
+            Return to frontpage
+          </Button>
+        }
+        icon={<BugReportIcon fontSize="medium" />}
+      >
+        {props.message || 'Unknown error occurred'}
+      </CenteredAlert>
+    );
 
-    unauthorizedApi()
-      .externalTokenLogin({ externalToken })
-      .then((token) => {
-        if (token.externalTokenLogin && !token.externalTokenLogin.rejection) {
-          handleLogin(token.externalTokenLogin.token);
-          window.location.href = '/';
-        } else {
-          if (externalAuthLoginUrl) {
-            window.location.href = externalAuthLoginUrl;
-          }
+    const LoadingMessage = () => (
+      <CenteredAlert
+        severity="info"
+        action={
+          <Button
+            color="inherit"
+            size="small"
+            variant="outlined"
+            onClick={() => history.push('/')}
+          >
+            Cancel
+          </Button>
+        }
+        icon={<Lock />}
+      >
+        <AnimatedEllipsis>Please wait</AnimatedEllipsis>
+      </CenteredAlert>
+    );
+
+    const ContactingAuthorizationServerMessage = () => (
+      <CenteredAlert
+        severity="info"
+        action={
+          <Button
+            color="inherit"
+            size="small"
+            variant="outlined"
+            onClick={() => history.push('/')}
+          >
+            Cancel
+          </Button>
         }
-      });
+        icon={<Lock fontSize="medium" />}
+      >
+        <AnimatedEllipsis>Contacting authorization server</AnimatedEllipsis>
+      </CenteredAlert>
+    );
+
+    const handleCode = (code: string) => {
+      const { protocol, host, pathname } = window.location;
+      const currentUrlWithoutParams = [protocol, '//', host, pathname].join('');
+
+      setView(<ContactingAuthorizationServerMessage />);
+
+      unauthorizedApi()
+        .externalTokenLogin({
+          externalToken: code,
```

Comment:
Given that it seems `code` is only used once, in `handleCode` to set `externalToken`, could we rename it to `token` through this file? I think that'd be easier to follow, as eading `code` my mind keeps expecting to mean some javascript or an http status code, and there's no comments saying what `code` should be

Reply:
Yes, name `code` is quiet generic. I will change it to `authorizationCode`. The token would be a bad name for OAuth flow, because `token` in OAuth is a different thing. `code` short lived string that is is exchanged for for `token`. check this image https://darutk.medium.com/diagrams-and-movies-of-all-the-oauth-2-0-flows-194f3c3ade85 step 6.